Samples/BackgroundBlur/BackgroundBlur/BackgroundBlur.cpp

```diff
@@ -0,0 +1,515 @@
+        RECT rc;
+        GetClientRect(hwnd, &rc);
+        FillRect(hdc, &rc, (HBRUSH)COLOR_WINDOW);
+    }
+    EndPaint(hwnd, &ps);
+}
+
+//  Handler for WM_SIZE messages.
+void OnResize(WORD width, WORD height)
+{
+    if (g_pPlayer)
+    {
+        g_pPlayer->ResizeVideo(width, height);
+    }
+}
+
+
+// Handler for WM_CHAR messages. 
+void OnKeyPress(WPARAM key)
+{
+    switch (key)
+    {
+        // Space key toggles between running and paused
+    case VK_SPACE:
+        if (g_pPlayer->GetState() == Started)
+        {
+            g_pPlayer->Pause();
+        }
+        else if (g_pPlayer->GetState() == Paused)
+        {
+            g_pPlayer->Play();
+        }
+        break;
+    }
+}
+
+// Handler for Media Session events.
+void OnPlayerEvent(HWND hwnd, WPARAM pUnkPtr)
+{
+    HRESULT hr = g_pPlayer->HandleEvent(pUnkPtr);
+    if (FAILED(hr))
+    {
+        NotifyError(hwnd, L"An error occurred.", hr);
+    }
+    UpdateUI(hwnd, g_pPlayer->GetState());
+}
+
+
+// Update the application UI to reflect the current state.
+
+void UpdateUI(HWND hwnd, PlayerState state)
+{
+    BOOL bWaiting = FALSE;
+    BOOL bPlayback = FALSE;
+
+    assert(g_pPlayer != NULL);
+
+    switch (state)
+    {
+    case OpenPending:
+        bWaiting = TRUE;
+        break;
+
+    case Started:
+        bPlayback = TRUE;
+        break;
+
+    case Paused:
+        bPlayback = TRUE;
+        break;
+    }
+
+    HMENU hMenu = GetMenu(hwnd);
+    UINT  uEnable = MF_BYCOMMAND | (bWaiting ? MF_GRAYED : MF_ENABLED);
+
+    EnableMenuItem(hMenu, ID_FILE_OPENFILE, uEnable);
+    EnableMenuItem(hMenu, ID_FILE_OPENURL, uEnable);
+
+    if (bPlayback && g_pPlayer->HasVideo())
+    {
+        g_bRepaintClient = FALSE;
+    }
+    else
+    {
+        g_bRepaintClient = TRUE;
+    }
+}
+
+//  Show a message box with an error message.
+void NotifyError(HWND hwnd, PCWSTR pszErrorMessage, HRESULT hrErr)
+{
+    const size_t MESSAGE_LEN = 512;
+    WCHAR message[MESSAGE_LEN];
+
+    if (SUCCEEDED(StringCchPrintf(message, MESSAGE_LEN, L"%s (HRESULT = 0x%X)",
+        pszErrorMessage, hrErr)))
+    {
+        MessageBox(hwnd, message, NULL, MB_OK | MB_ICONERROR);
+    }
+}
+
+
+//  Dialog proc for the "Open URL" dialog.
+INT_PTR CALLBACK OpenUrlDialogProc(HWND hDlg, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    static OpenUrlDialogInfo* pUrl = NULL;
+
+    BOOL result = FALSE;
+
+    switch (message)
+    {
+    case WM_INITDIALOG:
+        // The caller sends a pointer to an OpenUrlDialogInfo structure as the 
+        // lParam. This structure stores the URL.
+        pUrl = (OpenUrlDialogInfo*)lParam;
+        return (INT_PTR)TRUE;
+
+    case WM_COMMAND:
+        switch (LOWORD(wParam))
+        {
+        case IDOK:
+            if (pUrl)
+            {
+                // Get the URL from the edit box in the dialog. This function 
+                // allocates memory. The caller must call CoTaskMemAlloc.
```

Comment:
Comment looks misleading (caller must call CoTaskMemFree, right?). Also it seems to be leaking the allocation, I think you'd have to check if pUrl->pszURL is not null and free it before calling AllocGetWindowText.

Reply:
is this what you mean? 

`if (pUrl->pszURL) 
                {
                    std::free(pUrl->pszURL);
                }`

Reply:
`CoTaskMemFree(pUrl->pszURL);` in the caller (the OnOpenURL function)